File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    
    stack = deque()
    truck_weights = deque(truck_weights)
    count = 0
    
    # while문 돌면서 count 증가
    while truck_weights or stack:
        count+=1 # time
        
        # 나갈 트럭
        if any(count - i[1] >= bridge_length for i in stack):
            stack.popleft()
        
        # length랑 sum체크해서 더들어갈 수 있으면
        if truck_weights and sum(i[0] for i in stack)+truck_weights[0] <= weight and\
        len(stack) + 1 <= bridge_length:
            # 튜플 형태로 넣기
            stack.append((truck_weights.popleft(), count))
        
            
    return count

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
// truck_weights의 길이는 1 이상 10,000 이하
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int time = 0;
        Queue<Integer> bridge = new LinkedList<>();
        Queue<Integer> truck = new LinkedList<>();
        int curWeight = 0;
        int curLength = 0;
        int count = 0;
        int N = truck_weights.length;
        while (count < N) {
            // 다리에서 나가는 트럭
            if (!bridge.isEmpty() && bridge.peek() + bridge_length == time) {
                bridge.poll();
                int w = truck.poll();
                curWeight -= w;
                curLength -= 1;
            }
            
            int w = truck_weights[count];
            if (curWeight + w <= weight && curLength + 1 <= bridge_length) {
                curWeight += w;
                curLength += 1;
                bridge.add(time);
                truck.add(w);
                count++;
            }
            time++;
        }
        
        return time + bridge_length;
    }
}